File: code/python/studentfilter.py
#!/usr/bin/env python

class Student:
    def __init__(self, firstname, lastname, finalexamscore):
        self.firstname      = firstname
        self.lastname       = lastname
        self.finalexamscore = finalexamscore

    def has_passed(self):
        return self.finalexamscore >= 60

def passed(students):
    passed_students = []
    for student in students:
        if student.has_passed():
            passed_students.append(student)
    return passed_students

students = [Student("John", "Deer", 60),
            Student("Billy", "Bob", 49.1),
            Student("Jane", "Doe", 89),
            Student("Jack", "Johnson", 29.3)]

if __name__ == '__main__':
    print  "Students that have passed:\n", passed(students)

File: code/haskell/studentfilter.hs
data Student = Student { firstName      :: String
                       , lastName       :: String
                       , finalExamScore :: Double}

instance Show Student where
         show (Student firstName lastName score) = unwords [firstName, lastName, "("++ show score ++ ")"]

passed :: [Student] -> [Student]
passed students = filter has_passed students

has_passed :: Student -> Bool
has_passed student = finalExamScore student >= 60

students = [Student {firstName="John",  lastName="Deer",    finalExamScore=60},
            Student {firstName="Billy", lastName="Bob",     finalExamScore=49.1},
            Student {firstName="Jane",  lastName="Doe",     finalExamScore=89},
            Student {firstName="Jack",  lastName="Johnson", finalExamScore=29.3}]

main :: IO()
main = do
    putStrLn $ "Students that have passed:\n" ++ show (passed students)
